// @microsoft/sp-dialog/lib/FabricDialogWrapper.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
var React = require("react");
var office_ui_fabric_react_bundle_1 = require("@microsoft/office-ui-fabric-react-bundle");
var sp_webpart_base_1 = require("@microsoft/sp-webpart-base");
var DialogManager_1 = require("./DialogManager");
var FabricDialogWrapper = (function (_super) {
    __extends(FabricDialogWrapper, _super);
    function FabricDialogWrapper(props) {
        var _this = _super.call(this) || this;
        _this._onDismissed = _this._onDismissed.bind(_this);
        _this.state = {
            isOpen: false,
            isContentReady: false
        };
        return _this;
    }
    FabricDialogWrapper.prototype.render = function () {
        var _this = this;
        if (this.state.dialog) {
            return (React.createElement(office_ui_fabric_react_bundle_1.workaround_Modal, { isOpen: this.state.isOpen, isBlocking: this.state.dialog._configuration.isBlocking, onDismissed: this._onDismissed, onDismiss: this._closeCallback },
                React.createElement("div", { ref: function (c) { return _this._contentDiv = c; } }),
                !this.state.isContentReady &&
                    React.createElement(office_ui_fabric_react_bundle_1.DialogContent, null,
                        React.createElement(sp_webpart_base_1.PlaceholderSpinner, { label: '' }))));
        }
        else {
            return React.createElement("div", null);
        }
    };
    FabricDialogWrapper.prototype.open = function (dialog) {
        var _this = this;
        this._openPromise = new Promise(function (res, rej) {
            _this._resolveOpenPromise = res;
        });
        dialog = dialog || this.state.dialog;
        this.setState({ dialog: dialog, isOpen: true, isContentReady: false });
        return this._openPromise;
    };
    FabricDialogWrapper.prototype.close = function () {
        var _this = this;
        this._closedPromise = new Promise(function (res, rej) {
            _this._resolveClosedPromise = res;
        });
        if (!this.state.isOpen) {
            this._resolveClosedPromise();
        }
        else {
            this.setState({ isOpen: false });
        }
        return this._closedPromise;
    };
    FabricDialogWrapper.prototype.componentWillMount = function () {
        this._detachedElement = document.createElement('DIV');
    };
    FabricDialogWrapper.prototype.componentDidUpdate = function () {
        var _this = this;
        if (this.state.isOpen && this.state.dialog && this._detachedElement && this._contentDiv) {
            if (!this.state.isContentReady) {
                this._detachedElement.innerHTML = '';
                DialogManager_1._getInternalDialogApi(this.state.dialog)._render(this._detachedElement).then(function () {
                    _this.setState({ isContentReady: true });
                });
            }
            else {
                this._contentDiv.innerHTML = '';
                this._contentDiv.appendChild(this._detachedElement);
                this._resolveOpenPromise();
            }
        }
    };
    FabricDialogWrapper.prototype._onDismissed = function () {
        if (!this.state.isOpen && this._resolveClosedPromise) {
            this._resolveClosedPromise();
        }
    };
    FabricDialogWrapper.prototype._closeCallback = function () {
        if (this.props.closeCallback && this.state.dialog) {
            this.props.closeCallback(this.state.dialog);
        }
    };
    __decorate([
        office_ui_fabric_react_bundle_1.autobind
    ], FabricDialogWrapper.prototype, "render", null);
    __decorate([
        office_ui_fabric_react_bundle_1.autobind
    ], FabricDialogWrapper.prototype, "open", null);
    __decorate([
        office_ui_fabric_react_bundle_1.autobind
    ], FabricDialogWrapper.prototype, "close", null);
    __decorate([
        office_ui_fabric_react_bundle_1.autobind
    ], FabricDialogWrapper.prototype, "componentDidUpdate", null);
    __decorate([
        office_ui_fabric_react_bundle_1.autobind
    ], FabricDialogWrapper.prototype, "_onDismissed", null);
    __decorate([
        office_ui_fabric_react_bundle_1.autobind
    ], FabricDialogWrapper.prototype, "_closeCallback", null);
    return FabricDialogWrapper;
}(React.Component));
exports.default = FabricDialogWrapper;
